File: src/engine/heuristic-scorer.ts
import { SealIssue, Verdict, verdictFromScore } from '../types.js'

// Below this, the deterministic layer isn't confident enough to call it a clean PASS on its
// own (matches scoreToVerdict's PASS threshold in types.ts) — used by `--llm auto` (AD-5) to
// decide whether escalating to the LLM reviewer is worth it. Conservative on purpose: escalate
// more readily rather than less (openspec/changes/verifiable-gate-hardening/intent.md Risks).
export const AMBIGUOUS_THRESHOLD = 85

export class HeuristicScorer {
  static computeDetectorScore(findings: Array<{ trust_deduction?: number }>): number {
    const total = findings.reduce((sum, f) => sum + (f.trust_deduction ?? 0), 0)
    return Math.max(0, Math.min(100, Math.floor(100 - total)))
  }

  static computeFinalScore(detectorScore: number, policyDeductions: number): number {
    return Math.max(0, Math.min(100, Math.floor(detectorScore - policyDeductions)))
  }

  static toVerdict(score: number, blockingIssues: SealIssue[]): Verdict {
    return verdictFromScore(score, blockingIssues.length > 0)
  }

  static isAmbiguous(score: number): boolean {
    return score < AMBIGUOUS_THRESHOLD
  }
}

File: python/seal_gate/engine/heuristic_scorer.py
import math
from ..types import SealIssue, VERDICT_ORDER

# Mirrors src/engine/heuristic-scorer.ts AMBIGUOUS_THRESHOLD — used by Seal.with_llm(mode='auto').
AMBIGUOUS_THRESHOLD = 85


def is_ambiguous(score: int) -> bool:
    return score < AMBIGUOUS_THRESHOLD


def verdict_from_score(score: int, has_blocking: bool) -> str:
    base = score_to_verdict(score)
    if has_blocking:
        idx = max(VERDICT_ORDER.index(base), VERDICT_ORDER.index('REVISE'))
        return VERDICT_ORDER[idx]
    return base


def score_to_verdict(score: int) -> str:
    if score >= 85: return 'PASS'
    if score >= 70: return 'PASS_WITH_WARNINGS'
    if score >= 50: return 'REVISE'
    if score >= 30: return 'ESCALATE_TO_HUMAN'
    return 'BLOCK'


def compute_detector_score(findings: list) -> int:
    total = sum(getattr(f, 'trust_deduction', 0) or 0 for f in findings)
    return max(0, min(100, math.floor(100 - total)))


def compute_final_score(detector_score: int, policy_deductions: int) -> int:
    return max(0, min(100, math.floor(detector_score - policy_deductions)))
